crates/stateless/src/validation.rs

```diff
@@ -171,16 +212,241 @@ pub fn stateless_validation_with_trie<T, ChainSpec, E>(
     witness: ExecutionWitness,
     chain_spec: Arc<ChainSpec>,
     evm_config: E,
-) -> Result<B256, StatelessValidationError>
+) -> Result<(B256, HashedPostState), StatelessValidationError>
 where
     T: StatelessTrie,
     ChainSpec: Send + Sync + EthChainSpec<Header = Header> + EthereumHardforks + Debug,
     E: ConfigureEvm<Primitives = EthPrimitives> + Clone + 'static,
 {
-    let current_block = recover_block_with_public_keys(current_block, public_keys, &*chain_spec)?;
+    let ExecutionContext { block, parent, ancestor_hashes } = prepare_execution_context(
+        current_block,
+        public_keys,
+        &witness.headers,
+        chain_spec.clone(),
+    )?;
+
+    // Verify that the pre-state reads are correct
+    let (mut trie, bytecode) =
+        track_cycles!("verify_witness", T::new(&witness, parent.state_root)?);
+    let db = WitnessDatabase::new(&trie, bytecode, ancestor_hashes);
+
+    let hashed_post_state =
+        stateless_validation_execution(&block, &parent, chain_spec, evm_config, db)?;
```

Comment:
In short, the only thing that `stateleless_validation_execution` cares is receiving a proper `db`. If that under the hood is a `WitnessDatabase` or a `CacheDB<ExtDB>` (i.e., flatdb), is irrelevant.